x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/InvalidateApiKeyRequest.java

```diff
@@ -0,0 +1,134 @@
+    private String apiKeyId;
+    private String apiKeyName;
+
+    public InvalidateApiKeyRequest() {
+    }
+
+    public InvalidateApiKeyRequest(StreamInput in) throws IOException {
+        super(in);
+        realmName = in.readOptionalString();
+        userName = in.readOptionalString();
+        apiKeyId = in.readOptionalString();
+        apiKeyName = in.readOptionalString();
+    }
+
+    private InvalidateApiKeyRequest(@Nullable String realmName, @Nullable String userName, @Nullable String apiKeyId,
+            @Nullable String apiKeyName) {
+        this.realmName = realmName;
+        this.userName = userName;
+        this.apiKeyId = apiKeyId;
+        this.apiKeyName = apiKeyName;
+    }
+
+    public String getRealmName() {
+        return realmName;
+    }
+
+    public String getUserName() {
+        return userName;
+    }
+
+    public String getApiKeyId() {
+        return apiKeyId;
+    }
+
+    public String getApiKeyName() {
+        return apiKeyName;
+    }
+
+    /**
+     * Creates invalidate api key request for given realm name
+     * @param realmName realm name
+     * @return {@link InvalidateApiKeyRequest}
+     */
+    public static InvalidateApiKeyRequest usingRealmName(String realmName) {
+        return new InvalidateApiKeyRequest(realmName, null, null, null);
+    }
+
+    /**
+     * Creates invalidate API key request for given user name
+     * @param userName user name
+     * @return {@link InvalidateApiKeyRequest}
+     */
+    public static InvalidateApiKeyRequest usingUserName(String userName) {
+        return new InvalidateApiKeyRequest(null, userName, null, null);
+    }
+
+    /**
+     * Creates invalidate API key request for given realm and user name
+     * @param realmName realm name
+     * @param userName user name
+     * @return {@link InvalidateApiKeyRequest}
+     */
+    public static InvalidateApiKeyRequest usingRealmAndUserName(String realmName, String userName) {
+        return new InvalidateApiKeyRequest(realmName, userName, null, null);
+    }
+
+    /**
+     * Creates invalidate API key request for given api key id
+     * @param apiKeyId api key id
+     * @return {@link InvalidateApiKeyRequest}
+     */
+    public static InvalidateApiKeyRequest usingApiKeyId(String apiKeyId) {
+        return new InvalidateApiKeyRequest(null, null, apiKeyId, null);
+    }
+
+    /**
+     * Creates invalidate api key request for given api key name
+     * @param apiKeyName api key name
+     * @return {@link InvalidateApiKeyRequest}
+     */
+    public static InvalidateApiKeyRequest usingApiKeyName(String apiKeyName) {
+        return new InvalidateApiKeyRequest(null, null, null, apiKeyName);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (Strings.hasText(realmName) == false && Strings.hasText(userName) == false && Strings.hasText(apiKeyId) == false
```

Comment:
also there are other issues here. Api key id and name will be ignored if realm or username is specified. I think you need to look at the transport action and ensure the various cases are covered here with validation